receiver/opentelemetryreceiver/opentelemetry.go

```diff
@@ -0,0 +1,270 @@
+// Copyright 2020, OpenTelemetry Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package opentelemetryreceiver
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"net"
+	"net/http"
+	"sync"
+
+	gatewayruntime "github.com/grpc-ecosystem/grpc-gateway/runtime"
+	collectormetricspb "github.com/open-telemetry/opentelemetry-proto/gen/go/collector/metrics/v1"
+	collectortracepb "github.com/open-telemetry/opentelemetry-proto/gen/go/collector/traces/v1"
+	"github.com/rs/cors"
+	"github.com/soheilhy/cmux"
+	"google.golang.org/grpc"
+
+	"github.com/open-telemetry/opentelemetry-collector/component"
+	"github.com/open-telemetry/opentelemetry-collector/consumer"
+	"github.com/open-telemetry/opentelemetry-collector/observability"
+	"github.com/open-telemetry/opentelemetry-collector/oterr"
+	"github.com/open-telemetry/opentelemetry-collector/receiver"
+	"github.com/open-telemetry/opentelemetry-collector/receiver/opentelemetryreceiver/metrics"
+	"github.com/open-telemetry/opentelemetry-collector/receiver/opentelemetryreceiver/trace"
+)
+
+// Receiver is the type that exposes Trace and Metrics reception.
+type Receiver struct {
+	mu                sync.Mutex
+	ln                net.Listener
+	serverGRPC        *grpc.Server
+	serverHTTP        *http.Server
+	gatewayMux        *gatewayruntime.ServeMux
+	corsOrigins       []string
+	grpcServerOptions []grpc.ServerOption
+
+	traceReceiverOpts   []trace.Option
+	metricsReceiverOpts []metrics.Option
+
+	traceReceiver   *trace.Receiver
+	metricsReceiver *metrics.Receiver
+
+	traceConsumer   consumer.TraceConsumer
+	metricsConsumer consumer.MetricsConsumer
+
+	stopOnce                 sync.Once
+	startServerOnce          sync.Once
+	startTraceReceiverOnce   sync.Once
+	startMetricsReceiverOnce sync.Once
+}
+
+var _ receiver.MetricsReceiver = (*Receiver)(nil)
+var _ receiver.TraceReceiver = (*Receiver)(nil)
+
+// New just creates the OpenTelemetry receiver services. It is the caller's
+// responsibility to invoke the respective Start*Reception methods as well
+// as the various Stop*Reception methods to end it.
+func New(transport string, addr string, tc consumer.TraceConsumer, mc consumer.MetricsConsumer, opts ...Option) (*Receiver, error) {
+	// TODO: (@odeke-em) use options to enable address binding changes.
```

Comment:
Suggest to remove any original OC author's name and assign them unofficial tasks :)

Reply:
> Let's add a log warning to the factory that says this is meant for development purposes only and not intended to be used for real use cases.

We should make this production quality before Mar 16 Beta release. OK if it is not ready in the first PR and we have a warning but we will need to make it ready quickly.